cloudant-sync-datastore-core/src/main/java/com/cloudant/sync/replication/GetRevisionTaskThreaded.java

```diff
@@ -137,14 +142,36 @@ public boolean add(Future<DocumentRevsList> documentRevsListFuture) {
         // we make the request at construction time...
         for (final BulkGetRequest request : requests) {
 
-            completionService.submit(new Callable<DocumentRevsList>() {
+            submittedJobs.add(completionService.submit(new Callable<DocumentRevsList>() {
                 @Override
                 public DocumentRevsList call() throws Exception {
                     return new DocumentRevsList(GetRevisionTaskThreaded.this.sourceDb.getRevisions
                             (request.id, request.revs, request.atts_since,
                                     GetRevisionTaskThreaded.this.pullAttachmentsInline));
                 }
-            });
+            }));
+        }
+    }
+
+    void cancel() {
+        synchronized (submittedJobs) {
+            for (Future<DocumentRevsList> f : submittedJobs) {
+                // Don't interrupt if already running, we'll get the running ones from the completed
+                // queue.
+                f.cancel(false);
+            }
+            // Cancelling will line a whole bunch of tasks up for addition to the queue, so we need
+            // to keep draining it
+            try {
+                while (!submittedJobs.isEmpty()) {
+                    Future<DocumentRevsList> r = completionService.take();
```

Comment:
As you mentioned, I would be a bit wary of a race condition creeping in between submittedJobs and  completionService - I wonder if there are any circumstances under which `take()` could block because there's nothing on the completionService? Maybe easier to `poll()` then `take()`?